go/tools/release-notes/release_notes.go

```diff
@@ -0,0 +1,264 @@
+		execErr := err.(*exec.ExitError)
+		return prInfo{}, fmt.Errorf("%s:\nstderr: %s\nstdout: %s", err.Error(), execErr.Stderr, out)
+	}
+	var prInfo prInfo
+	err = json.Unmarshal(out, &prInfo)
+	return prInfo, err
+}
+
+func loadAllPRs(prs []string) ([]prInfo, error) {
+	errChan := make(chan error)
+	wgDone := make(chan bool)
+	prChan := make(chan string, len(prs))
+	// fill the work queue
+	for _, s := range prs {
+		prChan <- s
+	}
+	close(prChan)
+
+	var prInfos []prInfo
+	fmt.Printf("Found %d merged PRs. Loading PR info", len(prs))
+	wg := sync.WaitGroup{}
+	mu := sync.Mutex{}
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			// load meta data about PRs
+			defer wg.Done()
+			for b := range prChan {
+				fmt.Print(".")
+				prInfo, err := loadPRinfo(b)
+				if err != nil {
+					errChan <- err
+					break
+				}
+				mu.Lock()
+				prInfos = append(prInfos, prInfo)
+				mu.Unlock()
+			}
+		}()
+	}
+
+	go func() {
+		// wait for the loading to finish
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	var err error
+	select {
+	case <-wgDone:
+		break
+	case err = <-errChan:
+		break
+	}
+
+	fmt.Println()
+	return prInfos, err
+}
+
+func groupPRs(prInfos []prInfo) map[string]map[string][]prInfo {
+	prPerType := map[string]map[string][]prInfo{}
+
+	for _, info := range prInfos {
+		var typ, component string
+		for _, lbl := range info.Labels {
+			switch {
+			case strings.HasPrefix(lbl.Name, prefixType):
+				typ = strings.TrimPrefix(lbl.Name, prefixType)
+			case strings.HasPrefix(lbl.Name, prefixComponent):
+				component = strings.TrimPrefix(lbl.Name, prefixComponent)
+			}
+		}
+		switch typ {
+		case "":
+			typ = "Other"
+		case "Bug":
+			typ = "Bug fixes"
+		}
+
+		if typ == "" {
+			typ = "Other"
+		}
+		if component == "" {
+			component = "Other"
+		}
+		components, exists := prPerType[typ]
+		if !exists {
+			components = map[string][]prInfo{}
+			prPerType[typ] = components
+		}
+
+		prsPerComponentAndType := components[component]
+		components[component] = append(prsPerComponentAndType, info)
+	}
+	return prPerType
+}
+
+func createSortedPrTypeSlice(prPerType map[string]map[string][]prInfo) []prType {
+	var data []prType
+	for typeKey, typeElem := range prPerType {
+		newPrType := prType{
+			Name: typeKey,
+		}
+		for componentKey, prInfos := range typeElem {
+			newComponent := prComponent{
+				Name:    componentKey,
+				PrInfos: prInfos,
+			}
+			sort.Slice(newComponent.PrInfos, func(i, j int) bool {
+				return newComponent.PrInfos[i].Number < newComponent.PrInfos[j].Number
+			})
+			newPrType.Components = append(newPrType.Components, newComponent)
+		}
+		sort.Slice(newPrType.Components, func(i, j int) bool {
+			return newPrType.Components[i].Name < newPrType.Components[j].Name
+		})
+		data = append(data, newPrType)
+	}
+	sort.Slice(data, func(i, j int) bool {
+		return data[i].Name < data[j].Name
+	})
+	return data
+}
+
+func writePrInfos(fileout string, prPerType map[string]map[string][]prInfo) (err error) {
```

Comment:
> map[string]map[string][]prInfo

Can we create cascading `tyepdef` declarations so that we have something like `PullRequests`, `PullRequestLabels` or whatever it is that is implied by these `map`s?